File: backend/app/llm/ollama.py
import httpx
from .base import BaseLLM

OLLAMA_BASE = "http://localhost:11434"


class OllamaLLM(BaseLLM):
    def __init__(self, model: str = "llama3"):
        self.model = model

    @property
    def provider_name(self) -> str:
        return "ollama"

    async def list_models(self) -> list[str]:
        try:
            async with httpx.AsyncClient(timeout=3.0) as client:
                resp = await client.get(f"{OLLAMA_BASE}/api/tags")
                resp.raise_for_status()
                data = resp.json()
                return [m["name"] for m in data.get("models", [])]
        except Exception:
            return []

    async def chat(
        self,
        messages: list[dict],
        temperature: float = 0.7,
        max_tokens: int = 2048,
    ) -> str:
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": False,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
            },
        }
        async with httpx.AsyncClient(timeout=120.0) as client:
            resp = await client.post(f"{OLLAMA_BASE}/api/chat", json=payload)
            resp.raise_for_status()
            data = resp.json()
            return data.get("message", {}).get("content", "")

    async def chat_stream(
        self,
        messages: list[dict],
        temperature: float = 0.7,
        max_tokens: int = 2048,
    ):
        """Async generator that yields text chunks as they stream."""
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": True,
            "options": {
                "temperature": temperature,
                "num_predict": max_tokens,
            },
        }
        import json
        async with httpx.AsyncClient(timeout=120.0) as client:
            async with client.stream("POST", f"{OLLAMA_BASE}/api/chat", json=payload) as resp:
                resp.raise_for_status()
                async for line in resp.aiter_lines():
                    if line.strip():
                        try:
                            chunk = json.loads(line)
                            content = chunk.get("message", {}).get("content", "")
                            if content:
                                yield content
                        except Exception:
                            pass

File: src/services/ollama.js
// Ollama REST API service
// Docs: https://github.com/ollama/ollama/blob/main/docs/api.md

const BASE = 'http://localhost:11434'

/**
 * Fetch list of locally available models from Ollama.
 * Returns an array of model name strings, or [] if Ollama isn't running.
 */
export async function listModels() {
    try {
        const res = await fetch(`${BASE}/api/tags`, { signal: AbortSignal.timeout(2000) })
        if (!res.ok) return []
        const data = await res.json()
        return (data.models || []).map(m => m.name)
    } catch {
        return []
    }
}

/**
 * Check if Ollama is reachable.
 */
export async function ping() {
    try {
        await fetch(`${BASE}/api/tags`, { signal: AbortSignal.timeout(1500) })
        return true
    } catch {
        return false
    }
}

/**
 * Send a chat request to Ollama (streaming).
 * @param {string} model  - model name, e.g. "llama3"
 * @param {Array}  messages - OpenAI-style [{role, content}]
 * @param {Function} onChunk - called with each text chunk as it streams
 * @param {AbortSignal} signal - optional AbortSignal to cancel
 */
export async function chatStream(model, messages, onChunk, signal) {
    const res = await fetch(`${BASE}/api/chat`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model, messages, stream: true }),
        signal,
    })
    if (!res.ok) throw new Error(`Ollama error: ${res.status}`)

    const reader = res.body.getReader()
    const decoder = new TextDecoder()
    while (true) {
        const { done, value } = await reader.read()
        if (done) break
        const lines = decoder.decode(value).split('\n').filter(Boolean)
        for (const line of lines) {
            try {
                const json = JSON.parse(line)
                if (json.message?.content) onChunk(json.message.content)
            } catch { /* skip malformed chunks */ }
        }
    }
}

/**
 * Simple one-shot generate (no streaming).
 */
export async function generate(model, prompt) {
    const res = await fetch(`${BASE}/api/generate`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ model, prompt, stream: false }),
    })
    if (!res.ok) throw new Error(`Ollama error: ${res.status}`)
    const data = await res.json()
    return data.response
}
